### DISCUSSION TITLE
Best Practices for Initializing PocketBase Instance (Next.js)

### DISCUSSION THREAD

**Discussion by guilhermeguidetti** [pocketbase/pocketbase#5359] [Q&A] [answered]
Hi there,

I'm looking for guidance on how to properly initialize the PocketBase instance in a Next.js project. What would be the recommended approach for this? Should I be using `const pb = new PocketBase(process.env.POCKETBASE_URL)` in every page or function where I need to make calls to PocketBase, or is there a better practice for this?

Thanks in advance for your help!

**Comment by ganigeorgiev** [comment]
_Just a side-note:_ I personally haven't used Next.js in a very long time and I can't help you with best practice tips, but if you haven't already please make sure to check https://github.com/pocketbase/pocketbase/discussions/5313. It may seems strange coming from the author of the project, but you might not need PocketBase if you are going to use Next.js in SSR mode because you already have a working backend (the Next.js server). Alternatively if you still want to use PocketBase, the easiest and safest way to do that is to interact with it directly from the client-side (ex. SPA; see the footnote in the linked discussion). Integrating PocketBase with another backend (JS SSR framework) comes with a lot of caveats (_and Next.js due to its streaming oriented rendering design does't  make it easier, at least the last time I've checked_) and I'd suggest to reconsider your tooling/stack before investing too much time in building some needlessly overcomplicated setup.

**Reply by guilhermeguidetti** [reply]
I see, so would you recommend using Vite + React? I'm considering it since we already have a mobile project running with PocketBase, and now we're going to create the web version.

**Reply by ganigeorgiev** [reply, marked as answer]
> so would you recommend using Vite + React?

Yes, because a SPA would be easier, more performant and you are less likely to mess up the security, especially since you already have another client-side application (the mobile project) communicating directly with PocketBase.

**Reply by guilhermeguidetti** [reply]
Btw, now that I'll be using Vite, how do you recommend initializing PocketBase? Should I create a context for it?

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand what do you mean by "_create a context_" but when interacting with PocketBase from the client-side it is safe to have a single global/shared JS SDK instance, aka. you can simply import an exported instance.

You could check a Svelte + Vite example in https://github.com/presentator/presentator/tree/master/ui (_the shared PocketBase client is created in the [`src/pb.js` file](https://github.com/presentator/presentator/blob/master/ui/src/pb.js) and it is a little more complicated due to the multi auth collections support in the specific project; one example how the instance is used could be found in [PageProject.svelte](https://github.com/presentator/presentator/blob/master/ui/src/components/PageProject.svelte)_).

**Reply by guilhermeguidetti** [reply]
I'll take a look! That was exactly what I needed. Thanks!

**Comment by niallobrien** [comment]
Some thoughts here:
1. Nextjs can output a static site (page content fetched at build time) which can be great if you have a simple static website (eg. Marketing website) but want to manage the content in PocketBase. This solution is great if SEO is a primary concern.
2. The static site could be an "app" ie. a web app behind a login-wall. If so, you could use Next to build an entirely clientside app (single page app) and it would just build a static site so long as you don't use server functions or React server components.
3. Fullstack Next option with SSR - there's still some value here imo. You may have a dynamic website in which SEO is important but so is client side interactivity eg. e-commerce site. Being able to run additional logic on your Next server is also very useful and you could still use PocketBase's SDK in your server functions to fetch/update data on the server. Though PocketBase seems to be aimed more at utilising its SDK in the browser, you can definitely use it on the server if you so wish.

To answer your initial question, I'd import and initialise the SDK in `_app.tsx` so it's already there for each route.
